Spigot-Server-Patches/0140-Cache-uuid-requests-in-memory.patch

```diff
@@ -0,0 +1,732 @@
+ 
+     }
+ 
++    private PlayerProfile profile; // Paper - store bukkit profile (if pre-login fired)
+     public void disconnect(String s) {
++        // Paper start - fire internal login cancelled if pre login fired
++        if (profile != null) {
++            InternalListener.LISTENERS.forEach((listener) -> listener.onCancelledLogin(profile));
++        }
++        // Paper end
+         try {
+             LoginListener.c.info("Disconnecting " + this.d() + ": " + s);
+             ChatComponentText chatcomponenttext = new ChatComponentText(s);
+@@ -244,6 +255,11 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+                             java.util.UUID uniqueId = i.getId();
+                             final org.bukkit.craftbukkit.CraftServer server = LoginListener.this.server.server;
+ 
++                            // Paper start - fire pre join for all our internal listeners
++                            LoginListener.this.profile = ProfileUtils.toPaperWithProperties(i); // Mark the pre login event as 'fired', so cancel event will fire if needed
++                            InternalListener.LISTENERS.forEach((listener) -> listener.onAsyncPreLogin(profile));
++                            // Paper end
++
+                             AsyncPlayerPreLoginEvent asyncEvent = new AsyncPlayerPreLoginEvent(playerName, address, uniqueId);
+                             server.getPluginManager().callEvent(asyncEvent);
+ 
+diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
+index b1e6e5d..2462eb0 100644
+--- a/src/main/java/net/minecraft/server/MinecraftServer.java
++++ b/src/main/java/net/minecraft/server/MinecraftServer.java
+@@ -45,9 +45,10 @@ import org.bukkit.craftbukkit.CraftServer;
+ 
+ // CraftBukkit start
+ // CraftBukkit end
+-import co.aikar.timings.SpigotTimings; // Paper
+-
+ // Paper start
++import co.aikar.timings.SpigotTimings;
++import com.destroystokyo.paper.profile.BukkitGameProfileRepository;
++import com.destroystokyo.paper.profile.BukkitSessionService;
+ import java.util.LinkedList;
+ import java.util.Queue;
+ // Paper end
+@@ -565,7 +566,7 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
+         }
+     }
+     // Paper End
+- 
++
+     public void run() {
+         try {
+             if (this.init()) {
+@@ -1515,11 +1516,11 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
+     }
+ 
+     public MinecraftSessionService ay() {
+-        return this.V;
++        return new BukkitSessionService(this.V, Bukkit.getProfileLookup()); // PaperSpigot - return our sesion service
+     }
+ 
+     public GameProfileRepository getGameProfileRepository() {
+-        return this.W;
++        `return new BukkitGameProfileRepository(Bukkit.getProfileLookup()); // PaperSpigot - return our game profile reop
```

Comment:
Why are you returning a new profile repository each time this is called? Cache it.